lib/ukstore/include/uk/store.h

```diff
@@ -0,0 +1,452 @@
+	} get;
+
+	/* Function setter pointer */
+	union {
+		uk_store_set_s8_func_t   s8;
+		uk_store_set_u8_func_t   u8;
+		uk_store_set_s16_func_t  s16;
+		uk_store_set_u16_func_t  u16;
+		uk_store_set_s32_func_t  s32;
+		uk_store_set_u32_func_t  u32;
+		uk_store_set_s64_func_t  s64;
+		uk_store_set_u64_func_t  u64;
+		uk_store_set_uptr_func_t uptr;
+	} set;
+
+	/* The entry name */
+	const char *entry_name;
+
+	/* Entry flags */
+	__u16 flags;
+
+	/* Entry getter/setter type */
+	__u16 type;
+} __align8;
+
+struct uk_store_folder {
+	/* List for the dynamic folders only */
+	struct uk_list_head folder_head;
+
+	/* List for the entries in the folder */
+	struct uk_list_head entry_head;
+
+	/* The folder name */
+	const char *folder_name;
+} __align8;
+
+struct uk_store_folder_entry {
+	/* Saved entry pointer */
+	struct uk_store_entry *entry;
+
+	/* The linked list head */
+	struct uk_list_head list_head;
+
+	/* Refcount */
+	__atomic refcount;
+} __align8;
+
+#define UK_STORE_FLAG_STATIC	1
+
+/**
+ * Retuns the folder_entry of an entry
+ *
+ * @param entry_ptr the entry
+ */
+#define	uk_store_get_folder_entry(entry_ptr)	\
+	__containerof(entry_ptr, struct uk_store_folder_entry, entry)
+
+/* Static entry array start+end points */
+extern struct uk_store_entry *uk_store_entries_start;
+extern struct uk_store_entry *uk_store_entries_end;
+
+/* Do not call directly */
+#define __UK_STORE_INITREG_FOLDER(fldr, fldr_string)			\
+	static struct uk_store_folder					\
+	__used __section(".uk_store_lib_" fldr_string) __align8		\
+	__uk_store_folder_list ## _ ## fldr  = {			\
+		.entry_head = UK_LIST_HEAD_INIT(			\
+		__uk_store_folder_list ## _ ## fldr.entry_head),	\
+		.folder_name = STRINGIFY(fldr),				\
+		.folder_head = {NULL, NULL}				\
+	}
+
+/* Do not call directly */
+#define _UK_STORE_INITREG_FOLDER(fldr, fldr_string)	\
+	__UK_STORE_INITREG_FOLDER(fldr, fldr_string)
+
+/**
+ * Adds a folder to the folder section.
+ *
+ * @param fldr the folder to add
+ */
+#define UK_STORE_INITREG_FOLDER(fldr)	\
+	_UK_STORE_INITREG_FOLDER(fldr, STRINGIFY(__LIBNAME__))
+
+/**
+ * Adds an entry to the entry section.
+ *
+ * @param entry the entry in the section
+ * @param e_type the type of the function
+ * @param e_get getter pointer
+ * @param e_set setter pointer
+ */
+#define UK_STORE_INITREG_ENTRY(entry, e_type, e_get, e_set)		\
+	static const struct uk_store_entry				\
+	__used __section(".uk_store_entries_list") __align8		\
+	__uk_store_entries_list ## _ ## entry  = {			\
+		.entry_name = STRINGIFY(entry),				\
+		.type       = (e_type),					\
+		.get.e_type = (e_get),					\
+		.set.e_type = (e_set),					\
+		.flags      = UK_STORE_FLAG_STATIC			\
+	}
+
+/**
+ * Initializes an entry
+ *
+ * @param entry the entry to initialize
+ * @param e_name the name of the entry
+ * @param e_type the type of the entry
+ * @param getter getter pointer
+ * @param setter setter pointer
+ */
+#define uk_store_entry_init(entry, e_name, e_type, getter, setter)	\
```

Comment:
Hum... I am thinking in the meantime that the library usage is easier if we move memory management of the entries to the library itself. Encapsulating of the entries would be consistent (thus it is safer to use `containerof` internally) and registering code. Also, refcounting and releasing of memory gets easier.
In such a case, you would not have `init` functions anymore. What do you think?

I am thinking of an interface like this:
``` c
struct uk_folder *uk_store_folder_create(const char *name);

#define uk_store_dentry_create(folder, name, type, getter, setter) \
   uk_store_dentry_create_ ## type((folder), (name), (getter), (setter))

struct uk_entry_entry *uk_store_dentry_create_u8(struct uk_store_folder *f, uk_store_get_u8_func_t s, uk_store_set_u8_func_t f);
struct uk_entry_entry *uk_store_dentry_create_u16(struct uk_store_folder *f, uk_store_get_u16_func_t s, uk_store_set_u16_func_t f);
struct uk_entry_entry *uk_store_dentry_create_u32(struct uk_store_folder *f, uk_store_get_u32_func_t s, uk_store_set_u32_func_t f);
/* ... */
```

⚠️ One note regarding the name strings when creating folders/entries: Because an entry lives longer than the caller and we take over memory management of entries to the library, you should create a duplicate allocation for the name filed. Either you embedded it to the struct (like 8 or 16 chars max) or do a `strdup()` or `strndup()`. We should not expect that the caller keeps the string alive.

Reply:
Sure, sounds good, I'll do the edit now.

Reply:
It seems to me you are suggesting that the entry create function also creates the folder_entry. This is fine too.

Reply:
Almost done with this.